lib/quicly.c

```diff
@@ -3402,14 +3420,58 @@ static int handle_version_negotiation_packet(quicly_conn_t *conn, quicly_decoded
 #undef CAN_SELECT
 }
 
-int quicly_is_destination(quicly_conn_t *conn, int is_1rtt, ptls_iovec_t cid)
+static int compare_socket_address(struct sockaddr *x, struct sockaddr *y)
 {
-    if (quicly_cid_is_equal(&conn->super.host.cid, cid)) {
-        return 1;
-    } else if (!is_1rtt && !quicly_is_client(conn) && quicly_cid_is_equal(&conn->super.host.offered_cid, cid)) {
-        /* long header pacekt carrying the offered CID */
-        return 1;
+#define CMP(a, b)                                                                                                                  \
+    if (a != b)                                                                                                                    \
+    return a < b ? -1 : 1
+
+    CMP(x->sa_family, y->sa_family);
+
+    if (x->sa_family == AF_INET) {
+        struct sockaddr_in *xin = (void *)x, *yin = (void *)y;
+        CMP(ntohl(xin->sin_addr.s_addr), ntohl(yin->sin_addr.s_addr));
+        CMP(ntohs(xin->sin_port), ntohs(yin->sin_port));
+    } else if (x->sa_family == AF_INET6) {
+        struct sockaddr_in6 *xin6 = (void *)x, *yin6 = (void *)y;
+        int r = memcmp(xin6->sin6_addr.s6_addr, yin6->sin6_addr.s6_addr, sizeof(xin6->sin6_addr.s6_addr));
+        if (r != 0)
+            return r;
+        CMP(ntohs(xin6->sin6_port), ntohs(yin6->sin6_port));
+        CMP(xin6->sin6_flowinfo, yin6->sin6_flowinfo);
+        CMP(xin6->sin6_scope_id, yin6->sin6_scope_id);
+    } else {
+        assert(!"unknown sa_family");
     }
+
+#undef CMP
+    return 0;
+}
+
+int quicly_is_destination(quicly_conn_t *conn, struct sockaddr *sa, socklen_t salen, quicly_decoded_packet_t *decoded)
+{
+    if (QUICLY_PACKET_IS_LONG_HEADER(decoded->octets.base[0])) {
+        /* long header: validate address, then consult the CID */
+        if (compare_socket_address(conn->super.peer.sa, sa) != 0)
```

Comment:
Does this mean that we only accept incoming packets to the same socket that were used at connection establishment?  Is there any reason we shouldn't accept packets that might arrive on a different socket?

Reply:
No. This is checking the peer's (essentially client's) address.

My understanding is that client's 2-tuple is required consistent during the handshake (though I cannot find the corresponding text in the transport draft). This is necessary, to avoid seeing two client-generated CIDs arriving from different addresses considered as belonging to the same connection.